apiserver/facades/agent/uniter/networkinfo.go

```diff
@@ -422,20 +411,142 @@ func (n *NetworkInfo) NetworksForRelation(
+			// Remove this ID to prevent further processing.
+			spaceSet.Remove(corenetwork.AlphaSpaceId)
+		} else {
+			privateIPAddress = privateMachineAddress.Value
+		}
 	}
 
-	return machine.GetNetworkInfoForSpaces(set.NewStrings(spaces...)), nil
+	// Link-layer devices are set in a single transaction for all devices
+	// observed on the machine, so the first result will include them all.
+	var addresses []*state.Address
+	retryArg := n.retryFactory()
+	retryArg.Func = func() error {
+		var err error
+		addresses, err = machine.AllAddresses()
+		return err
+	}
+	retryArg.IsFatalError = func(err error) bool {
+		return err != nil
+	}
+	if err := retry.Call(retryArg); err != nil {
+		result := machineNetworkInfoResult{Error: errors.Annotate(err, "getting devices addresses")}
+		for _, id := range spaceSet.Values() {
+			if _, ok := results[id]; !ok {
+				results[id] = result
+			}
+		}
+		return results, nil
+	}
+
+	logger.Debugf("Looking for address from %v in spaces %v", addresses, spaceIDs)
+
+	var privateLinkLayerAddress *state.Address
+	for _, addr := range addresses {
+		subnet, err := addr.Subnet()
+		switch {
+		case errors.IsNotFound(err):
+			logger.Debugf("skipping %s: not linked to a known subnet (%v)", addr, err)
+
+			// For a space-less model, we will not have subnets populated,
+			// and will therefore not find a subnet for the address.
+			// Capture the link-layer information for machine private address
+			// so that we can return as much information as possible.
+			// TODO (manadart 2020-02-21): This will not be required once
+			// discovery (or population of subnets by other means) is
+			// introduced for the non-space IAAS providers (LXD, manual, etc).
+			if addr.Value() == privateIPAddress {
+				privateLinkLayerAddress = addr
+			}
+		case err != nil:
+			logger.Errorf("cannot get subnet for address %q - %q", addr, err)
+		default:
+			if spaceSet.Contains(subnet.SpaceID()) {
+				r := results[subnet.SpaceID()]
+				r.NetworkInfos, err = addAddressToResult(r.NetworkInfos, addr)
+				if err != nil {
+					r.Error = err
+				} else {
+					results[subnet.SpaceID()] = r
+				}
+			}
+
+			// TODO (manadart 2020-02-21): This reflects the behaviour prior
+			// to the introduction of the alpha space.
+			// It mimics the old behaviour for the empty space ("").
+			// If that was passed in, we included the machine's preferred
+			// local-cloud address no matter what space it was in,
+			// treating the request as space-agnostic.
+			// To preserve this behaviour, we return the address as a result
+			// in the alpha space no matter its *real* space if addresses in
+			// the alpha space were requested.
+			// This should be removed with the institution of universal mutable
+			// spaces.
+			if spaceSet.Contains(corenetwork.AlphaSpaceId) && addr.Value() == privateIPAddress {
+				r := results[corenetwork.AlphaSpaceId]
+				r.NetworkInfos, err = addAddressToResult(r.NetworkInfos, addr)
+				if err != nil {
+					r.Error = err
```

Comment:
Is it possible to enter both if blocks here?

If you entered the if block on L488 and got an error and you also get an error here, you could be potentially overwriting the previously captured error.

Reply:
I think this was something that @SimonRichardson brought up when this logic was moved from state (it is unchanged, just relocated). I didn't want to start changing it at that time.

We both agreed that the results accrual strategy used here should be simplified. That work will be undertaken in _develop_.